### PR TITLE
Change the code of the package to Typescript

### PR DESCRIPTION
**Pull request by khushal87**
With this PR the codebase for the Package is changed to Typescript and there has been also improved in terms of converting most of the components to use React hooks architecture, etc.
Changes with this PR:

- Changed the package codebase to **Typescript**
- Change the demo-app codebase to **Typescript**
- Fix all the **typescript errors/warning(around 80)** of both package and demo-app
- Change the ** Class-based components to Functional components**.
- Fix **Related issues**. Eg: linting, demo-app
- Added **Status bar in demo screens** so that the app looks better on android devices.

@sshivananda @Monte9 Please review. Thank you :)
Reference issue: #143

### CODE REVIEW

**Comment by sshivananda** [comment]
awesome work : going to test this out over the weekend. can you also take a look at the build failures when you get the chance

**Comment by khushal87** [comment]
Hey, @sshivananda if you go through the details there some issue of connection which is been shown and due to which the checks are failing. I am not sure why this is happening. Can you guide me?

**Comment by khushal87** [comment]
@Monte9 @sshivananda please review

File: src/index.d.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import * as React from "react";
-import PropTypes from "prop-types";
```
Comment (sshivananda) [inline src/index.d.ts:2]: is this file required?

Comment (khushal87) [reply src/index.d.ts:2]: No, this can be ignored too as we have the typescript exports now @sshivananda

File: package.json [hunk 2]
```diff
@@ -20,8 +20,8 @@
     "ui-components"
```
Comment (sshivananda) [inline package.json:20]: should line 4 be updated?

Comment (khushal87) [reply package.json:20]: Can you brief what do you want to say by this? Thanks beforehand @sshivananda

Comment (sshivananda) [reply package.json:20]: That was a very quick reply !!

I think lines 4 and 5 should be
```
"main": "./dist/src/index.js",
"types": "./dist/src/index.d.ts",
```

File: src/index.d.ts (hunk 1 above)
Comment (sshivananda) [reply src/index.d.ts:2]: yeah - I think we should delete this file 
there is an autogenerated dist/src/index.d.ts - that one should contain everything that is contained in this file

File: package.json (hunk 2 above)
Comment (khushal87) [reply package.json:20]: If you test the code without the changes I have made the lint won't have any input to apply upon. That's why it won't work. Therefore, I have added it. You can try it on your end too. @sshivananda

**Comment by sshivananda** [comment]
@khushal87 minor feedback 
also I think we should update packge.json for the demo folder as well

File: package.json [hunk 3]
```diff
@@ -20,8 +20,8 @@
...
   ],
   "scripts": {
-    "lint": "eslint src/",
-    "lint-demo": "eslint demo/src/",
+    "lint": "eslint --ext .ts --ext .tsx --ext .d.ts src/",
+    "lint-demo": "eslint --ext .ts --ext .tsx demo/src/",
```
Comment (sshivananda) [inline package.json:24]: can we also add a build script
```
"build": "tsc",
```

File: src/index.d.ts (hunk 1 above)
Comment (khushal87) [reply src/index.d.ts:2]: Should I delete it with a commit? @sshivananda

File: package.json (hunk 3 above)
Comment (khushal87) [reply package.json:24]: Sure, we can do it. @sshivananda . It will foster be for local typescript checks

**Comment by khushal87** [comment]
> @khushal87 minor feedback
> also I think we should update packge.json for the demo folder as well

With this?
```
"build": "tsc",
```
@sshivananda

File: package.json (hunk 2 above)
Comment (sshivananda) [reply package.json:20]: What did you mean by
```If you test the code without the changes I have made the lint won't have any input to apply upon. That's why it won't work. Therefore, I have added it. You can try it on your end too.```

Comment (khushal87) [reply package.json:20]: Nothing, sorry I misunderstood your statement. Will fix this. @sshivananda

File: src/index.d.ts (hunk 1 above)
Comment (sshivananda) [reply src/index.d.ts:2]: sure !

Comment (khushal87) [reply src/index.d.ts:2]: Only one doubt, what about the comments of the props, do you want me to add those comments in the individual files where the type is defined for the component or we can ignore them now? @sshivananda

Comment (sshivananda) [reply src/index.d.ts:2]: we can add them in the individual files

Comment (khushal87) [reply src/index.d.ts:2]: Okay

File: demo/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (sshivananda) [inline demo/tsconfig.json:1]: Adding new comment to continue conversation
```
Shiva: @khushal87 minor feedback
also I think we should update packge.json for the demo folder as well

Khushal:
With this?

"build": "tsc",
@sshivananda
```

Comment (sshivananda) [reply demo/tsconfig.json:1]: yeah - I think the build script would be useful 
apart from that I think we also need three dev dependencies
```
"@types/node": "^14.14.37",
 "@types/react-native": "~0.62.13",
"typescript": "^4.2.3"
```

**Comment by khushal87** [comment]
@sshivananda fixed and added the asked changes

**Review by sshivananda — APPROVED** [review]

**Comment by sshivananda** [comment]
Awesome work @khushal87

**Comment by sshivananda** [comment]
Released in v8.0.1

**Comment by khushal87** [comment]
Thank you @sshivananda
